### PR TITLE
Force mesh by polylines

### PR DESCRIPTION
**Pull request by vcloarec**
In the scope of mesh editing, this PR allows the user to force the mesh by existing polylines (as break lines). Forcing lines consist of line that the faces are forced to follow, that is edges of encountered faces have to be on theses lines.

forcing_line

The user choose vector layers geometries in the map canvas by selecting them. Then, the user clicks on the dedicated push button to force the mesh on the lines contains by these geometries.

The different options are:
- add new vertices when the lines cut the internal edges
- interpolate the Z value of new added vertices on the mesh or from the lines
- Tolerance : when a lines is at a distance from an existing vertex lower than the tolerance, no new vertex is created but the existing vertex is used

forceByBreakLine

**_Note about the terminology:_** 
These type of lines can be called break lines. The word "force" is use here because, usually, especially when TIN are constructed, break lines are permanent lines that will be used even later for new faces operations (add, remove flip, ..) to continue to constrained the mesh. In that case, break lines are part on the mesh data, and are identified as this. Here, we just apply this "forcing" line and leave, so for latter operation, the faces are not constrained anymore, as opposite as real break line. Maybe later we could have some real break lines that will permanent in mesh layer, so it is important to differentiate the terminology today.

### CODE REVIEW

**Comment by vcloarec** [comment]
again unrelated failure of `test_core_networkaccessmanager`

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 1]
```diff
@@ -1897,6 +1916,195 @@ QgsTopologicalMesh::Changes QgsTopologicalMesh::addVertexInface( int includingFa
   return changes;
 }
 
+QgsTopologicalMesh::Changes QgsTopologicalMesh::insertVertexInFacesEdge( int faceIndex, int position, const QgsMeshVertex &vertexToInsert )
+{
+  const QgsMeshFace face1 = mMesh->face( faceIndex );
+
+//  const int addedVertexIndex = mMesh->vertexCount();
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.cpp:1923]: remove commented out code

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 2]
```diff
@@ -1897,6 +1916,195 @@ QgsTopologicalMesh::Changes QgsTopologicalMesh::addVertexInface( int includingFa
...
+//  int newFaceIndexStartIndex = mMesh->faceCount();
+
+  Changes changes;
+  changes.mVerticesToAdd.append( vertexToInsert );
+  changes.mAddedFacesFirstIndex = mMesh->faceCount();
+  // triangulate the first face
+
+  int newVertexPositionInFace1 = position + 1;
+
+  auto triangulate = [this, &changes]( int removedFaceIndex, const QgsMeshVertex & newVertex, int newVertexPosition, QVector<int> &edgeFacesIndexes )->bool
```
Comment (PeterPetrik) [inline src/core/mesh/qgstopologicalmesh.cpp:1933]: can we make new function (static?) out of it for readability?

File: src/core/mesh/qgsmeshforcebypolylines.h [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+#include "qgsmeshadvancedediting.h"
+
+/**
+ * \ingroup core
+ *
+ * \brief Class derived from QgsMeshAdvancedEditing that forces mesh based on a line
+ *
+ * Caller of this class has to set the line with setInputLine() before applying the edition with QgsMeshEditor::advancedEdit()
+ *
+ * Other option has also to be et before calling QgsMeshEditor::advancedEdit()
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshforcebypolylines.h:33]: et -> set

File: src/core/mesh/qgsmeshadvancedediting.cpp [hunk 4]
```diff
@@ -773,3 +782,5 @@ QgsMeshVertex QgsMeshTransformVerticesByExpression::transformedVertex( QgsMeshLa
     return layer->nativeMesh()->vertex( vertexIndex );
 }
 
+
+
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshadvancedediting.cpp:786]: ?

File: src/app/mesh/qgsmaptooleditmeshframe.h [hunk 5]
```diff
@@ -60,6 +65,49 @@ class APP_EXPORT QgsZValueWidget : public QWidget
...
+    {
+      Mesh,
+      Lines
+    };
+    Q_ENUM( IntepolationMode )
+
+    //! Constructor
+    QgsMeshEditForceByLineAction( QObject *parent = nullptr );
+
+    //! Set the associated map canvas, used notably for map unit
```
Comment (PeterPetrik) [inline src/app/mesh/qgsmaptooleditmeshframe.h:83]: sets

**Comment by PeterPetrik** [comment]
Hi @vcloarec, thanks for the PR, I have few points before I do deep review:

1. can you please add an explanation to this PR description **what** is "force the mesh by existing polylines (as break lines)" by non-mesh people? Also please add a description of **how** to use it ( e.g.  I assume you have to first select line/polyline from vector layer ...)

2. maybe language problem, but I am confused by the word "forcing/force" as used in the classes description overall the code. Is it some standard terminology/word used for this functionality? at least we need to add descriptions to the new classes for future developers so they understand what the classes are for and what they intend to do.

3.  GUI-wise I am not sure if we have a push-button in the main toolbar with the advanced options anywhere else in the QGIS (@nirvn ?). I am not expert on gui, but one way how to do it would maybe be:
 a) in push down menu you should just select "mode" (==either interpolate Z value from mesh or forcing line) similarly to measure tool or other tools with different modes.
 b) in case tolerance and the checkbox are not changed frequently , they can be hidden similar way how you set up for example snapping settings?

File: src/core/mesh/qgstopologicalmesh.cpp (hunk 2 above)
Comment (vcloarec) [reply src/core/mesh/qgstopologicalmesh.cpp:1933]: not sure, with a static function I could not access to the private member of Changes class

**Comment by vcloarec** [comment]
>     3\. GUI-wise I am not sure if we have a push-button in the main toolbar with the advanced options anywhere else in the QGIS

@PeterPetrik , the settings in the push button is already used at least:

- Digitizing with trace: for the offset
- more recently, stream digitizing, for the streaming tolerance.
- 
Here, this implemented as the way as for the stream digitizing.

**Comment by nyalldawson** [comment]
Is this another candidate for exposing as a processing algorithm instead?

**Review by PeterPetrik — APPROVED** [review]

**Comment by PeterPetrik** [comment]
Let me merge this for 3.22, we will consider changes in GUI and adding it as processing alg. too. My understanding of the client's usage of this tool is that they need to use it interactively only, so we should keep the "button/toolbar" option too. We will try to think about refining the UX, so it is more consistent.

**Pull request merged**